routes/slots.js

```diff
@@ -1,34 +1,214 @@
+     slot: // conflict slot name
+     conflictGroup:// conflict slot group name
+   }
+   groupOption:
+ }
+ */
+slots.post('/addGroupValidate',auth.ensureAuthenticated, function (req, res) {
+  var passDataId = [];
+  var conflictDataName = [];
+  var count = 0;
+  // find slot groups
+  SlotGroup.find(function(err, groupOption) {
+    if (err) {
+      console.error(err);
+      return res.status(500).send(err.message);
+    }
+    // validate slot
+    Slot.find({
+      '_id': {$in: req.body.slotIds}
+    }, function (err, docs) {
+      if (err) {
+        console.error(err);
+        return res.status(500).send(err.message);
+      }
+      // divied two parts by inGroup field
+      var conflictData = [];
+      docs.forEach(function (d) {
+        if (d.inGroup) {
+          conflictData.push(d);
+        } else {
+          passDataId.push(d._id);
+        }
+      });
+
+      if(conflictData.length > 0) {
+        conflictData.forEach(function (r) {
+          SlotGroup.findOne({'_id': r.inGroup}, function(err, conflictGroup) {
+            conflictDataName.push({
+              slot: r.name,
+              conflictGroup: conflictGroup.name
+            });
+            count = count + 1;
+            if (count === conflictData.length) {
+              res.status(200).json({
+                passDataId: passDataId,
+                conflictDataName: conflictDataName,
+                groupOption: groupOption
+              });
+            }
+          });
+        });
+      }else {
+        res.status(200).json({
+          passDataId: passDataId,
+          conflictDataName: conflictDataName,
+          groupOption: groupOption
+        });
+      }
+    });
+  });
+});
+
+/*
+ return json data:
+ {
+   passDataId: // slot Ids can be removed
+   conflictDataName: {
+     slot: // conflict slot name
+   }
+ }
+ */
+slots.post('/removeGroupValidate',auth.ensureAuthenticated, function (req, res) {
+  var passDataId = [];
+  var conflictDataName = [];
+  var count = 0;
+  // validate slot
+  Slot.find({
+    '_id': {$in: req.body.slotIds}
+  }, function (err, docs) {
+    if (err) {
+      console.error(err);
+      return res.status(500).send(err.message);
+    }
+    docs.forEach(function (d) {
+      if (d.inGroup) {
+        passDataId.push(d._id);
+      } else {
+        conflictDataName.push({
+          slot: d.name
+        });
+      }
+      count = count + 1;
+      if (count === docs.length) {
+        res.status(200).json({
+          passDataId: passDataId,
+          conflictDataName: conflictDataName
+        });
+      }
+    });
+  });
+});
+
+slots.post('/addGroup',auth.ensureAuthenticated, function (req, res) {
+  // change slots of slotGroup
+  SlotGroup.findOne({'name': req.body.slotGroupName}, function(err, slotGroup){
+    if(err) {
+      console.error(err);
+      return res.status(500).send(err.message);
+    }
+    slotGroup.slots = slotGroup.slots.concat(req.body.slotIds);// concat array
```

Comment:
we need to do validation here. check if a slot id exists and the slot.inGroup is undefined.